File: Iteration-2/src/Course.java
import java.util.ArrayList;
import org.apache.log4j.Logger;

public abstract class Course {
    private Logger log = Logger.getLogger(Course.class);
    private String courseCode;
    private String courseName;
    private int courseCredit;
    private Schedule courseSchedule;
    private int quota;
    private Instructor courseInstructor;
    private ArrayList<Student> students;
    private int quotaProblem;
    private int collisionProblem;
    private int failedCredits;
    private int failedPreq;

    public Course(String courseName, String courseCode, int courseCredit, int courseDay, String courseHour, int quota) {
        this.courseName = courseName;
        this.courseCode = courseCode;
        this.courseCredit = courseCredit;
        this.courseSchedule = new Schedule(courseDay, courseHour);
        this.quota = quota;
        this.students = new ArrayList<Student>();
        this.quotaProblem = 0;
        this.collisionProblem = 0;
        this.failedCredits = 0;
        this.failedPreq = 0;
        log.info(this.getCourseName() + " (" + this.getCourseCode() + ")" + " named course created.");
    }

    public abstract boolean isEligibleToRequest(Student student);

    // GETTER & SETTER
    public String getCourseName() {
        return courseName;
    }

    public String getCourseCode() {
        return courseCode;
    }

    public int getQuato() {
        return quota;
    }

    public Instructor getCourseInstructor() {
        return courseInstructor;
    }

    public void setCourseInstructor(Instructor courseInstructor) {
        this.courseInstructor = courseInstructor;
        /*
         * log.info(this.getCourseName() + " now has a instructor: " +
         * this.getCourseInstructor());
         */
    }

    public ArrayList<Student> getStudents() {
        return this.students;
    }

    public int getCourseCredit() {
        return courseCredit;
    }

    public Schedule getCourseSchedule() {
        return this.courseSchedule;
    }

    public int getQuotaProblem() {
        return this.quotaProblem;
    }

    public void setQuotaProblem(int quotaProblem) {
        this.quotaProblem = quotaProblem;
        log.info(this.getCourseName() + ": Number of quota problem changed." + "(" + "New: " + this.getQuotaProblem()
                + ")");
    }

    public int getCollisionProblem() {
        return this.collisionProblem;
    }

    public void setCollisionProblem(int collisionProblem) {
        this.collisionProblem = collisionProblem;
        log.info(
                this.getCourseName() + ": Number of collision problem changed." + "(" + "New: "
                        + this.getCollisionProblem()
                        + ")");
    }

    public int getFailedCredits() {
        return this.failedCredits;
    }

    public void setFailedCredits(int failedCredits) {
        this.failedCredits = failedCredits;
        log.info(this.getCourseName() + ": Number of failed credits changed." + "(" + "New: " + this.getFailedCredits()
                + ")");
    }

    public int getFailedPreq() {
        return this.failedPreq;
    }

    public void setFailedPreq(int failedPreq) {
        this.failedPreq = failedPreq;
        log.info(this.getCourseName() + ": Number of failed prerequisite changed." + "(" + "New: "
                + this.getFailedPreq()
                + ")");
    }
}

File: Iteration-3/Code/Course.py
from abc import ABC, abstractmethod
import logging
from Schedule import Schedule


class Course(ABC):

    def __init__(self, courseName, courseCode, courseCredit, courseDay, courseHour, quota):
        self.__courseName = courseName
        self.__courseCode = courseCode
        self.__courseCredit = courseCredit
        self.__courseSchedule = Schedule(courseDay, courseHour)
        self.__quota = quota
        self.__students = []
        self.__quotaProblem = 0
        self.__collisionProblem = 0
        self.__failedCredits = 0
        self.__failedPreq = 0
        self.__enrolledCourseLimit = 0
        self.__courseInstructor = None
        logging.info(self.getCourseName() + " (" + self.getCourseCode() + ")" +
                     " named course created.")

    @abstractmethod
    def isEligibleToRequest(self, student):
        pass

    def getCourseName(self):
        return self.__courseName

    def getCourseCode(self) -> str:
        return self.__courseCode

    def getQuato(self):
        return self.__quota

    def getCourseInstructor(self):
        return self.__courseInstructor

    def setCourseInstructor(self, courseInstructor):
        self.__courseInstructor = courseInstructor

    def getStudents(self):
        return self.__students

    def getCourseCredit(self):
        return self.__courseCredit

    def getCourseSchedule(self):
        return self.__courseSchedule

    def getQuotaProblem(self):
        return self.__quotaProblem

    def setQuotaProblem(self, quotaProblem):
        self.__quotaProblem = quotaProblem
        logging.info(self.getCourseName() + ": Number of quota problem changed." +
                     "(" + "New: " + str(self.getQuotaProblem()) + ")")

    def getCollisionProblem(self):
        return self.__collisionProblem

    def setCollisionProblem(self, collisionProblem):
        self.__collisionProblem = collisionProblem
        logging.info(self.getCourseName() + ": Number of collision problem changed." +
                     "(" + "New: " + str(self.getCollisionProblem()) + ")")

    def getFailedCredits(self):
        return self.__failedCredits

    def setFailedCredits(self, failedCredits):
        self.__failedCredits = failedCredits
        logging.info(self.getCourseName() + ": Number of failed credits changed." +
                     "(" + "New: " + str(self.getFailedCredits()) + ")")

    def getFailedPreq(self):
        return self.__failedPreq

    def setFailedPreq(self, failedPreq):
        self.__failedPreq = failedPreq
        logging.info(self.getCourseName() + ": Number of failed prerequisite changed." +
                     "(" + "New: " + str(self.getFailedPreq()) + ")")

    def getEnrolledCourseLimit(self):
        return self.__enrolledCourseLimit

    def setEnrolledCourseLimit(self, courseLimit):
        self.__enrolledCourseLimit = courseLimit
        logging.info(self.getCourseName() + ": Number of course limit changed." +
                     "(" + "New: " + str(self.getEnrolledCourseLimit()) + ")")